### ISSUE TITLE
Invalid argument (22, EINVAL) because ioctl request or ioctl data is not valid

### ISSUE THREAD

**Issue by grantcurell** [luigirizzo/netmap#562]
I'm investigating an issue in tcpreplay, but am not sure if that issue might actually be in netmap. See: https://github.com/appneta/tcpreplay/issues/527#issuecomment-438383687.

I've got everything running in GDB. My arguments to that function are:

sp->handle.fd
```

(gdb) print sp->handle.fd
$11 = 3
```

NIOCREGIF (kinda obvious on that one ;-D)

**nmr**
```
(gdb) print nmr
$10 = {nr_name = "enp130s0f0\000\000\000\000\000", nr_version = 13, nr_offset = 0, nr_memsize = 0, nr_tx_slots = 0, nr_rx_slots = 0, nr_tx_rings = 0,
  nr_rx_rings = 0, nr_ringid = 0, nr_cmd = 0, nr_arg1 = 0, nr_arg2 = 0, nr_arg3 = 0, nr_flags = 1, spare2 = {0}}
```

I'm still reading about what all these things are, but the file handle appears valid. You can see from the source here: https://github.com/appneta/tcpreplay/blob/master/src/common/netmap.c that it is properly set to /dev/netmap. The problem line in questions is 343. That leads me to believe it would have to be a problem with the contents of nmr, but I'm still figuring out what all those fields mean. Am I missing anything a more experienced observer sees?

**Comment by grantcurell** [comment]
Tried netmap's example program and got the below:

```
[root@trafficgenerator nmreplay]# ./nmreplay -f /root/test_pcap.pcap -i enp130s0f0
752.408 main       [ 1247]      running on cores 2 8 4 10
752.409 readpcap   [  326]      filesize is 380318128
752.409 readpcap   [  344]      magic is 0xa1b2c3d4
752.409 readpcap   [  367]      swap 0 res 1000

752.411 readpcap   [  386]      --WWW-- reordered packet 10159

752.411 readpcap   [  386]      --WWW-- reordered packet 10734

752.414 readpcap   [  386]      --WWW-- reordered packet 22125

752.416 readpcap   [  386]      --WWW-- reordered packet 30851

752.424 readpcap   [  386]      --WWW-- reordered packet 72650

752.424 readpcap   [  386]      --WWW-- reordered packet 73742

752.425 readpcap   [  386]      --WWW-- reordered packet 75427

752.429 readpcap   [  386]      --WWW-- reordered packet 97123

752.432 readpcap   [  386]      --WWW-- reordered packet 112734

752.437 readpcap   [  386]      --WWW-- reordered packet 143570

752.444 readpcap   [  386]      --WWW-- reordered packet 182911

752.444 readpcap   [  386]      --WWW-- reordered packet 188167

752.453 readpcap   [  386]      --WWW-- reordered packet 242785

752.459 readpcap   [  386]      --WWW-- reordered packet 283683

752.465 readpcap   [  386]      --WWW-- reordered packet 319215

752.465 readpcap   [  402]      tot_pkt 319353 tot_bytes 375208456 tx_time 2143.241079 s first_len 54
752.465 readpcap   [  417]      first_pkt_time 0.000308 s
752.465 pcap_prod  [  780]      --- start create 319353 packets at tail 0
752.465 pcap_prod  [  804]      insert     0 len 54 cur_tt 0.000308
752.726 pcap_prod  [  836]      done q->prod_tail:391617792
752.735 nmreplay_m [  942]      --EEE-- cannot open netmap on enp130s0f0
753.409202 main [1276] exiting on abort
```

The NIC in question is an Intel x710:

```
lshw

        *-network:0
             description: Ethernet interface
             product: Ethernet Controller X710 for 10GbE SFP+
             vendor: Intel Corporation
             physical id: 0
             bus info: pci@0000:82:00.0
             logical name: enp130s0f0
             version: 01
             serial: 3c:fd:fe:b2:4c:00
             size: 10Gbit/s
             capacity: 10Gbit/s
             width: 64 bits
             clock: 33MHz
             capabilities: pm msi msix pciexpress vpd bus_master cap_list rom ethernet physical fibre 1000bt-fd 10000bt-fd autonegotiation
             configuration: autonegotiation=off broadcast=yes driver=i40e driverversion=2.4.6 duplex=full firmware=5.05 0x800029eb 1.1313.0 ip=169.254.232.64
             resources: irq:128 memory:fb000000-fb7fffff memory:fb818000-fb81ffff memory:fbe80000-fbefffff memory:fb820000-fb89ffff

```

**Comment by fklassen** [comment]
As per latest comments in https://github.com/appneta/tcpreplay/issues/527 ... this appears a bug in netmap master branch 11aa43a6. It was working for me a couple weeks ago, and it works if I checkout tag:v11.4.

Request: can you make sure you create branches for experimental code and keep master production ready? This is the standard for most GitHub projects.

**Comment by vmaffione** [comment]
@grantcurrell what does dmesg say?
This looks like an API mismatch, due to a recent API bump. Have you tried to recompile and install both netmap and your application?

The nmr is correct with all zero and only the interface name specified. Most of the parameters are in/out, so if you specify zero it means "no preference", and on return you will have them filled with the actual values.

@fklassen : we don't merge experimental code in master if it can break existing users.
The tags were created for specific reasons/users: they are not meant to be releases, nor to be used. We don't have releases, but we use a rolling release approach.

We have unit tests and integration tests to protect from regressions:
https://travis-ci.org/netmap-unipi/netmap
including several tests for the netmap API (https://github.com/luigirizzo/netmap/blob/master/utils/ctrl-api-test.c).

**Comment by grantcurell** [comment]
> @grantcurrell what does dmesg say?
> This looks like an API mismatch, due to a recent API bump. Have you tried to recompile and install both netmap and your application?
> 
> The nmr is correct with all zero and only the interface name specified. Most of the parameters are in/out, so if you specify zero it means "no preference", and on return you will have them filled with the actual values.
> 
> @fklassen : we don't merge experimental code in master if it can break existing users.
> The tags were created for specific reasons/users: they are not meant to be releases, nor to be used. We don't have releases, but we use a rolling release approach.
> 
> We have unit tests and integration tests to protect from regressions:
> https://travis-ci.org/netmap-unipi/netmap
> including several tests for the netmap API (https://github.com/luigirizzo/netmap/blob/master/utils/ctrl-api-test.c).

@vmaffione 

Unfortunately, in a desperate move to to see if I can get things working I blew away the Fedora Server 28 installation I had and replaced it with CentOS 7. Unfortunately, while I have very much enjoyed learning about how kernel drivers work I'm starting to butt up against my deadline hahaha. The server is the same box that was running Federoa Server 28, but am faced with a new error: the i40e driver in the configuration script says it has problems and isn't being compiled so at the moment I'm working through that. 

![image](https://user-images.githubusercontent.com/35169230/48625499-0dfe1180-e975-11e8-9d00-f18b4d5afb37.png)

I'm now deciding whether I want to troubleshoot this issue or move back to Fedora Server 28 where I previously had everything working before the stoppage. Nothing about the box has changed physically and the i40e driver compiled just fine on Fedora. Just sat down to see if I can figure it out now.

![image](https://user-images.githubusercontent.com/35169230/48625616-62a18c80-e975-11e8-8c24-b13332284782.png)

**Comment by grantcurell** [comment]
@vmaffione -opened a new bug: https://github.com/luigirizzo/netmap/issues/563

**Comment by vmaffione** [comment]
I would strongly recommend to follow these steps on your previously functioning Fedora 28 server.

````
$ cd /path/to/netmap/repo/
$ git clean -fdx # to clean the repository
$ git pull
$ ./configure [YOUR ARGUMENTS HERE]
$ make
$ sudo make install
$ # restart your applications or reboot
````
You can run the unit tests to make sure things are working
````
$ cd /path/to/netmap/repo
$ sudo modprobe netmap
$ sudo make unitest
````

**Comment by grantcurell** [comment]
> ```
> $ cd /path/to/netmap/repo
> ```

I'll try that! Right now I'm compiling CentOS' full kernel source. I've never done this before... holy crap it takes a while hahaha

**Comment by vmaffione** [comment]
The problems you see with i40 is that official Intel drivers do not support your kernel (it's too old).
You can try to build it against the kernel sources, providing the `--kernel-sources` and `--no-ext-drivers=i40e` to your `./configure` invocation.

**Comment by fklassen** [comment]
Thanks Vincenzo. I'll take back my comments. Everything works for me as long as I rebuild and reinstall everything.

I followed your procedures and `sudo make unitest` passed. Running `pkt-gen` also worked.

After recompiling `tcpreplay`, it also started to work.

It appears that when netmap updates, we must recompile. There is no backward compatibility.

**Comment by grantcurell** [comment]
@fklassen what OS are you running?

**Comment by fklassen** [comment]
Debian Stretch with 4.14 kernel. I had to install `libelf-dev` to compile.

**Comment by grantcurell** [comment]
> Debian Stretch with 4.14 kernel. I had to install `libelf-dev` to compile.

Thanks! I'll give that a shot!

**Comment by grantcurell** [comment]
> Thanks Vincenzo. I'll take back my comments. Everything works for me as long as I rebuild and reinstall everything.
> 
> I followed your procedures and `sudo make unitest` passed. Running `pkt-gen` also worked.
> 
> After recompiling `tcpreplay`, it also started to work.
> 
> It appears that when netmap updates, we must recompile. There is no backward compatibility.

I gave that a shot the other day on Fedora, but it was no dice for me for what it's worth. I deleted anything with the words netmap or tcpreplay from the box, recompiled netmap, loaded the i40e driver, then recompiled tcpreplay against the latest netmap sources and the error I originally had persisted. That's basically what lead me down the CentOS path.

**Comment by vmaffione** [comment]
@fklassen 
Sorry for the inconvenience, but it was necessary to update the ABI.

We actually provide backward compatibility: current applications (pkt-gen, bridge, lb, nmreplay, etc.) are still using the "old" ABI (NIOCREGIF). The new ABI is based on NIOCCTRL and is still not documented. If something broke (when using the old ABI), that was not intentional, and we will investigate that.

**Comment by vmaffione** [comment]
What changed was the (mostly unused) new ABI, and that change was done an a backward incompatible manner: but this was not meant to be an issue, because nobody is using the new ABI yet.
Note that the old ABI didn't change, so applications (that use the old ABI) should not have broken.
In any case, the change was needed to fix a mistake; we won't need to do something like that again.

**Comment by vmaffione** [comment]
For the sake of completeness I extended the test suite to test the legacy ABI (the one we are all using), with tests on forward and backward compatibility:
https://github.com/luigirizzo/netmap/blob/master/utils/ctrl-api-test.c#L186-L287
I don't see any failure, so I don't understand what went wrong with your environment.
In any case we should be covered now.

**Comment by grantcurell** [comment]
Confirmed. I reinstalled Fedora Server 29. My problems were a combination of some ignorance, my Intel X710 not accepting 3rd party NICs (jerks), and not prefixing the interface in nmreplay with "netmap:" which lead me to believe a test rig I had wasn't working when in fact it was... which then lead me on that whole CentOS journey.

But hey I ended up buying Linux Device Driver Development off of Amazon and man did I learn a mountain from this one hahaha.
